crates/ty_python_semantic/src/types/infer.rs

```diff
@@ -8853,6 +8870,50 @@ impl<'db, 'ast> TypeInferenceBuilder<'db, 'ast> {
         value_ty: Type<'db>,
         generic_class: ClassLiteral<'db>,
         generic_context: GenericContext<'db>,
+    ) -> Type<'db> {
+        let db = self.db();
+        let specialize = |types: &[Option<Type<'db>>]| {
+            Type::from(generic_class.apply_specialization(db, |_| {
+                generic_context.specialize_partial(db, types.iter().copied())
+            }))
+        };
+
+        self.infer_explicit_callable_specialization(
+            subscript,
+            value_ty,
+            generic_context,
+            specialize,
+        )
+    }
+
+    fn infer_explicit_type_alias_specialization(
+        &mut self,
+        subscript: &ast::ExprSubscript,
+        value_ty: Type<'db>,
+        generic_type_alias: PEP695TypeAliasType<'db>,
+        generic_context: GenericContext<'db>,
+    ) -> Type<'db> {
+        let db = self.db();
+        let specialize = |types: &[Option<Type<'db>>]| {
+            generic_type_alias.apply_specialization(db, |_| {
+                generic_context.specialize_partial(db, types.iter().copied())
+            })
+        };
+
+        self.infer_explicit_callable_specialization(
+            subscript,
+            value_ty,
+            generic_context,
+            specialize,
+        )
+    }
+
+    fn infer_explicit_callable_specialization(
+        &mut self,
+        subscript: &ast::ExprSubscript,
+        value_ty: Type<'db>,
+        generic_context: GenericContext<'db>,
+        specialize: impl FnOnce(&[Option<Type<'db>>]) -> Type<'db>,
```

Comment:
I think the "keep `GenericAlias` as a struct" suggestion would simplify this a bit, too. You'd end up with a `GenericAliasOrigin` enum, which wraps either a class literal or a type alias, and you could update `infer_explicit_class_specialization` to take in that instead of a `ClassLiteral`. That would eliminate the need for this callback (and the two new methods here), since the two callbacks would be combined into a new `GenericAliasOrigin::apply_specialization` method.